File: packages/protopy/protofy/agent/ProtoMqttAgent.py
import paho.mqtt.client as mqtt
import time
from .bifrost import gen_action_endpoint
from .ProtoAgentsInterface import ProtoAgentInterface
from .bifrost import register, pub_monitor, gen_action_endpoint


class ProtoMqttAgent(ProtoAgentInterface): 
    def __init__(self, agent_name): 
        self.name = agent_name
        self.client = mqtt.Client() 
        self.subsystems = []
        self.type = "mqtt"
        self.subsystems_handlers = {}

    def configure(self, subsystems) -> "ProtoMqttAgent": 
        self.subsystems = subsystems
        return self
         
    def connect(self, mqtt_host, mqtt_port): 
        self.client.connect(mqtt_host, mqtt_port) 
        self.client.loop_start()
        time.sleep(2)
        result = register(self.client.publish, self.name, self.subsystems)
        result.wait_for_publish()

        # consumer defined callback
        self.__consumer_callbacks_checker("on_connect")
        
    def pub_monitor(self, subsystem_name,  monitor_name, value): 
        result = pub_monitor(self.client.publish, self.name, subsystem_name, monitor_name, value) 
        result.wait_for_publish()

        # consumer defined callback
        self.__consumer_callbacks_checker("on_monitor_pub", value)

    def handle(self, subsystem_name, action_name, handler): 
        subsystem = next((s for s in self.subsystems if s["name"] == subsystem_name), None)
        if subsystem is None:
            raise ValueError(f"Subsystem '{subsystem_name}' not found.")
        
        action = next((a for a in subsystem.get("actions", []) if a["name"] == action_name), None)
        if action is None:
            raise ValueError(f"Action '{action_name}' not found in subsystem '{subsystem_name}'.")

        if subsystem_name not in self.subsystems_handlers:
            self.subsystems_handlers[subsystem_name] = {}

        self.subsystems_handlers[subsystem_name][action_name] = handler
        
        # register actions subscribers
        self.client.subscribe(gen_action_endpoint(self.name, subsystem_name, action_name))
        # register actions handlers
        self.__on_message()

        print(f"Handler assigned for action '{action_name}' in subsystem '{subsystem_name}'")


    def __on_message(self):
        self.client.on_message = on_message_factory(self.name, self.subsystems_handlers)

    def __consumer_callbacks_checker(self, name, value=None): 
        if hasattr(self, name) and callable(getattr(self, name)):
            callback = getattr(self, name) 
            if value is not None:
                callback(value)  
            else:
                callback()


def on_message_factory(agent_name, handlers): 
    def on_message(client, userdata, message): 
        payload = str(message.payload.decode("utf-8"))
        
        for subsystem_name, actions in handlers.items():
            for action_name, handler in actions.items():
                expected_topic = gen_action_endpoint(agent_name, subsystem_name, action_name)
                
                if message.topic == expected_topic:
                    handler(payload) 
                    return 

        print(f"No handler found for topic: {message.topic}")
        
        
    return on_message

File: packages/agents/js/ProtoMqttAgent.ts
import * as mqtt from 'mqtt';
import { genActionEndpoint, pubMonitor, register } from './bifrost';
import { ProtoAgentInterface } from './ProtoAgentsInterface';

export class ProtoMqttAgent extends ProtoAgentInterface {
    name: string
    client: mqtt.MqttClient
    listeners: any
    subsystems: Array<any>;
    subsystemsHandlers: { [key: string]: { [key: string]: (payload: any) => void } };
    type: string;

    constructor(agentName: string) {
        super()
        this.name = agentName;
        this.client = null as any;
        this.listeners = {} as any
        this.subsystems = [];
        this.type = "mqtt";
        this.subsystemsHandlers = {};
    }

    on(key: "connect" | "error" | "monitor_pub", cb: Function) {
        this.listeners[key] = cb
    }

    configure(subsystems) {
        this.subsystems = subsystems;
        return this;
    }

    connect(mqttHost, mqttPort, auth?: { username?: string, password?: string }) {
        const opts = (auth?.username && auth?.password) ? {
            username: auth.username,
            password: auth.password
        } : null;

        const url = `mqtt://${mqttHost}:${mqttPort}`;
        this.client = mqtt.connect(url, opts)
        try {
            this.client.on('error', (error) => {
                this.__consumerCallbacksChecker("error")
            });

            this.client.on('connect', () => {
                register(this.client.publish.bind(this.client), this.name, this.subsystems)
                this.__consumerCallbacksChecker("connect")
                // Register the message handler
                this.__onMessage();
            });

            return this
        } catch (err) {
            console.error("Cannot connect agent", err)
        }

        return null
    }

    pubMonitor(subsystemName, monitorName, value) {
        const subsystem = this.subsystems.find(s => s.name === subsystemName);
        if (!subsystem) {
            throw new Error(`Subsystem '${subsystemName}' not found.`);
        }

        const action = subsystem.monitors.find(a => a.name === monitorName);
        if (!action) {
            throw new Error(`Action '${monitorName}' not found in subsystem '${subsystemName}'.`);
        }

        pubMonitor(this.client.publish.bind(this.client), this.name, subsystemName, monitorName, value)

        // Consumer defined callback
        this.__consumerCallbacksChecker('monitor_pub', value);
    }

    handle(subsystemName, actionName, handler) {
        const subsystem = this.subsystems.find(s => s.name === subsystemName);
        if (!subsystem) {
            throw new Error(`Subsystem '${subsystemName}' not found.`);
        }

        const action = subsystem.actions.find(a => a.name === actionName);
        if (!action) {
            throw new Error(`Action '${actionName}' not found in subsystem '${subsystemName}'.`);
        }

        if (!this.subsystemsHandlers[subsystemName]) {
            this.subsystemsHandlers[subsystemName] = {};
        }

        this.subsystemsHandlers[subsystemName][actionName] = handler;

        // Register action subscriber
        const topic = genActionEndpoint(this.name, subsystemName, actionName);
        this.client.subscribe(topic, (err) => {
            if (err) {
                console.error(`Failed to subscribe to topic: ${topic}`, err);
            } else {
                console.log(`Subscribed to topic: ${topic}`);
            }
        });

        console.log(`Handler assigned for action '${actionName}' in subsystem '${subsystemName}'`);
    }

    __onMessage() {
        this.client.on('message', (topic, message) => {
            const payload = message.toString();
            const handlers = this.subsystemsHandlers;

            for (const [subsystemName, actions] of Object.entries(handlers)) {
                for (const [actionName, handler] of Object.entries(actions)) {
                    const expectedTopic = genActionEndpoint(this.name, subsystemName, actionName);
                    if (topic === expectedTopic) {
                        handler(payload);
                        return;
                    }
                }
            }

            console.log(`No handler found for topic: ${topic}`);
        });
    }

    __consumerCallbacksChecker(name, value = null) {
        if (typeof this.listeners[name] === 'function') {
            if (value !== null) {
                this.listeners[name](value);
            } else {
                this.listeners[name]();
            }
        }
    }
}